daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala

```diff
@@ -2364,6 +2361,35 @@ sealed abstract class DocumentPart(part: Node, parent: Document) {
     }
   }.trim.toUpperCase()
 
+  /*
+  * Description: Compare ByteDocumentPart/BitsDocumentPart string to a string of valid characters.
+  *              If the string runs into non-valid characters, skip over them.
+  *              If the string runs into "//", any character following it will be skipped (including valid ones).
+  *              Characters on a new line will be considered again, unless they are found after a "//"
+  * Parameters: String containing the acceptable characters (0,1 for bits) (0-9,A-F,a-f for bytes)
+  * Return: ListBuffer[Char] of the accepted characters from the ByteDocumentPart/BitsDocumentPart element
+  */
+  def commentCompatibleFormat(validCharacters: String) : ListBuffer[Char] = {
+    val tempListBuffer = new ListBuffer[Char]()
+    var noCommentBool = true
+
+    for (i <- 0 to partRawContent.length()-1){
```

Comment:
In general, iterating things with an index is frowned upon by many experienced programmers now, as it's too hard to tell if you've done the iteration just right or not. For someting that needs to be super-fast maybe. But for this usage clarity is way more important than speed. 

A way to do this that is much easier to review is more functional.

E.g., 
first split the string at line-endings to get a list of strings, one per line.
Then split each line at "//" discarding all but the first part.
Then replace any char matching regular expression "[^0-9a-fA-F]" with "". (for bytes case)
Then concatenate and mkstring from all the remaining pieces. 

Each of those operations should be 1 line of code.

Reply:
>  Then replace any char matching regular expression "[^0-9a-fA-F]" with "". (for bytes case)

Should we consider that an error rather than replacing invalid characters with empty string? Isn't the main purpose of the comment syntax to avoid cases where we accidentally misinterpret characters? I suggest any characters not in a comment *must* be a valid character for that the document type, or a space to allow for alignment/readability. Any other character should throw a TDMLException about an invalid document part.

Reply:
Yes, I agree with performing more strict validation of the document part contents (any invalid character except whitespace must be in a comment or must throw an exception).

Reply:
There are places where I know some non-whitespace separators are used for good effect E.g., things like

01|01 0001 10|11 

where the vertical bars are used to show where the byte boundaries or element boundaries are. So I think allowing whitespace and certain specific punctuation characters like brackets and parens in the mix is very useful. (at least '|()[].' are useful)

There is a fair amount of investment in old-style comments where people have done 

<documentPart type="bits"> trackStability 011 <documentPart>

So I think quite a lot of impact if that feature is removed allowing (and ignoring) those letters. 

Yes ultimately I would like to fix these, but many files (hundreds) of DFDL schema tests will have to be changed by users. 

I think we would need an option on the tdml:testSuite level to change the default of this from allowed to deprecated to disallowed.